sys/kern/subr_intr.c

```diff
@@ -1537,15 +1537,12 @@ dosoftints(void)
 void
 intr_pic_init_secondary(void)
 {
+	struct intr_pic *pic;
 
-	/*
-	 * QQQ: Only root PIC is aware of other CPUs ???
-	 */
-	KASSERT(intr_irq_root_dev != NULL, ("%s: no root attached", __func__));
-
-	//mtx_lock(&isrc_table_lock);
-	PIC_INIT_SECONDARY(intr_irq_root_dev);
-	//mtx_unlock(&isrc_table_lock);
+	//mtx_lock(&pic_list_lock);
+	STAILQ_FOREACH(pic, &pic_list, pic_next)
+		PIC_INIT_SECONDARY(pic->pic_dev);
```

Comment:
This isn't minimally required for arm64/Xen, as it changes the behaviour of other PIC drivers which implement PIC_INIT_SECONDARY.

Reply:
Have you noticed how many implementations of that there are?  The hook is barely used.  Most PICs leave the empty default.  As a result the hook is effectively broken for anything which isn't a root PIC.

Turning it into a root PIC _mildly_ abuses the functionality, but does sort of match what the device-tree looks like.

Reply:
Of course I noticed, otherwise I wouldn't have commented. The point is that there is more than zero, and I don't see any explanation in the commit logs of how they were tested, nor any explanation for why it is unlikely to cause problems.

Reply:
Being rarely used means it is simple to examine the uses and take actions believed to preserve functionality.  I lack any systems to _fully_ test this, but I can confirm it provides an appropriate hook to get Xen fully functional.

The default hook is a NOP.  Only GICv3 tries to propagate the call, but that requires a device-tree which conforms to the preconceptions of someone.  Since the one provide via Xen differs, there is a problem.

Anyway I was planning to abandon this approach as while the issue in the comment remains "QQQ: Only root PIC is aware of other CPUs ???".  Hooking in as a root also works.

I had hoped it would be part of #1363, but is now #1431.

Reply:
Similar to the above #1280 and #1363.  I'm still a bit unsure of how to handle this though.